### PR TITLE
Run travis on rustc-master instead of nightly

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -17,6 +17,11 @@ install:
     - rustup-init.exe -y --default-host %TARGET% --default-toolchain nightly
     - set PATH=%PATH%;C:\Users\appveyor\.cargo\bin;C:\Users\appveyor\.rustup\toolchains\nightly-%TARGET%\bin
     - if defined MSYS2_BITS set PATH=%PATH%;C:\msys64\mingw%MSYS2_BITS%\bin
+    - cargo install rustup-toolchain-install-master || echo "rustup-toolchain-install-master already installed"
+    - set RUSTC_HEAD=`git ls-remote https://github.com/rust-lang/rust.git HEAD | tr -d ",." | tr " \t" "\n" | grep -e "HEAD" -v`
```
Comment (Manishearth) [inline appveyor.yml]: let's use `rustup-toolchain-install-master -n master -f` from https://github.com/kennytm/rustup-toolchain-install-master/pull/6 when it merges

though I'm okay with merging this first and switching over later

**Comment by oli-obk** [comment]
hmm

```
+++travis_retry rustup-toolchain-install-master bc14d71622378cf942a834e7c2b5358b9901f775
./ci/base-tests.sh: line 4: travis_retry: command not found
```

**Comment by mati865** [comment]
Could `~/.cargo/bin` not be in PATH on Travis?

Anyway AppVeyor error looks worse:
```
= note: libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_end referenced in function _ZN4core3ptr13drop_in_place17h3973868bbd333ff7E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_easy_encoder referenced in function _ZN3xz26stream6Stream16new_easy_encoder17h444e67914679cfe5E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_alone_encoder referenced in function _ZN3xz26stream6Stream16new_lzma_encoder17h7a72f1d68ee5b478E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_stream_encoder referenced in function _ZN3xz26stream6Stream18new_stream_encoder17h698b6d2d519d1273E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_stream_decoder referenced in function _ZN3xz26stream6Stream18new_stream_decoder17hf2a8e53459d7da61E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_alone_decoder referenced in function _ZN3xz26stream6Stream16new_lzma_decoder17h2950602432ddcbd3E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_auto_decoder referenced in function _ZN3xz26stream6Stream16new_auto_decoder17hd4a704f64eb88947E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_code referenced in function _ZN3xz26stream6Stream7process17h416f627689d69b52E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_memlimit_get referenced in function _ZN3xz26stream6Stream8memlimit17hd44a326b1824d06eE
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_memlimit_set referenced in function _ZN3xz26stream6Stream12set_memlimit17h7f8ee5e5612bb876E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_lzma_preset referenced in function _ZN3xz26stream11LzmaOptions10new_preset17hb7cf472abb481456E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_check_is_supported referenced in function _ZN3xz26stream5Check12is_supported17h0c95786a9b51894dE
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_mf_is_supported referenced in function _ZN3xz26stream11MatchFinder12is_supported17h1143b95e3df085d9E
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_stream_encoder_mt_memusage referenced in function _ZN3xz26stream15MtStreamBuilder8memusage17h515e8f411dc5c4caE
          libxz2-2575134034ce30a1.rlib(xz2-2575134034ce30a1.xz26-c80f76e38c7a6925a75b74075834931c.rs.rcgu.o) : error LNK2019: unresolved external symbol lzma_stream_encoder_mt referenced in function _ZN3xz26stream15MtStreamBuilder7encoder17h4b074a2a8ad5634dE
          C:\Users\appveyor\AppData\Local\Temp\1\cargo-installqJxNyT\release\deps\rustup_toolchain_install_master-bb30363af798ef93.exe : fatal error LNK1120: 15 unresolved externals
          
error: aborting due to previous error
error:
 failed to compile `rustup-toolchain-install-master v1.1.0`, intermediate artifacts can be found at `C:\Users\appveyor\AppData\Local\Temp\1\cargo-installqJxNyT`
```

**Comment by oli-obk** [comment]
hmm... the macro imports aren't working.

**Comment by Manishearth** [comment]
Probably means it's not using the toolchain -- maybe it's obeying the
rust-toolchain file?

**Comment by Manishearth** [comment]
Rebased and updated to use new rustup-toolchain-install-master

**Comment by mati865** [comment]
Primary build should succeed if you apply this workaround https://github.com/rust-lang-nursery/rust-clippy/pull/3001
Other builds failed due to networking issue

**Comment by phansch** [comment]
I re-triggered the build after #3001 was merged. The master installer still sometimes fails with a Forbidden from the GitHub API. Perhaps we are hitting some rate limit?
```
fetching master commit hash... thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: ClientError(Forbidden), url: Some("https://api.github.com/repos/rust-lang/rust/commits/master") }
```

Looks like there are also different AppVeyor errors now:
```
= note: LINK : fatal error LNK1181: cannot open input file 'lzma.lib.lib'
```
and https://ci.appveyor.com/project/llogiq/rust-clippy/build/1.0.228#L265

**Comment by mati865** [comment]
@oli-obk `.lib` are static msvc libs, mingw does use `.a` extension for static lib.
Probably it's bug in the build system adding extension twice.

**Comment by mati865** [comment]
Please disregard my previous comment, I went too far with new component stuff. This repository needs to build Clippy everytime...

To debug this you can add following commands:
```
which rustc
rustc --print sysroot
ls $(rustc +master --print sysroot)/lib
```
It could be taking libs from the other toolchains.

**Comment by mati865** [comment]
My ["debug" build](https://travis-ci.org/mati865/rust-clippy/jobs/418403768):
```
$ which rustc
/home/travis/.cargo/bin/rustc
$ rustc --print sysroot
/home/travis/.rustup/toolchains/master
$ ls $(rustc +master --print sysroot)/lib
[..]
librustc_driver-b59ab39da03fc19f.so
[..]

/home/travis/.cargo/bin/clippy-driver: error while loading shared libraries: librustc_driver-b59ab39da03fc19f.so: cannot open shared object file: No such file or directory
```
Can we dump sysroot path from `clippy-driver`?

**Comment by flip1995** [comment]
The master-toolchain libs can be found at `$(rustc --print sysroot)/lib` after setting `rustup default master`. See https://github.com/flip1995/rust-clippy/blob/prs/ci/base-tests.sh#L15

**Comment by flip1995** [comment]
The appveyor linking error seems to be related with MinGW: https://github.com/alexcrichton/xz2-rs/commit/eceb2874389685814bfa68e81daa092ccf246383

I'm not sure why, but without MinGW the linking error disappears. But after that the dogfood test fails [appveyor](https://ci.appveyor.com/project/flip1995/rust-clippy#L685). I'm not sure if this is related to MinGW though..

I got travis to work, except sometimes
```bash
fetching master commit hash... thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: ClientError(Forbidden), url: Some("https://api.github.com/repos/rust-lang/rust/commits/master") }
```
https://github.com/flip1995/rust-clippy/commits/prs

Should I push this, to this branch?

**Comment by mati865** [comment]
Windows build system seems really borked then.

`Error { kind: ClientError(Forbidden), url: Some("https://api.github.com/repos/rust-lang/rust/commits/master") }`
Guess it's hitting GitHub request, is there any way to fetch last rust build hash and pass it across all Travis builds?
Anyway maybe @kennytm knows something about this error.

**Comment by kennytm** [comment]
GitHub will reject API requests from the CI since the IP is shared. You need to create a GitHub token. @mati865 

See https://github.com/kennytm/rustup-toolchain-install-master/blob/master/.travis.yml for an example (the `$GITHUB_TOKEN` is a secret env var which you could define in the Travis and AppVeyor settings)

**Comment by flip1995** [comment]
Thanks @kennytm!

Just tested it on my fork and it works like a charm! We now need someone to create the token for Clippy and add it to travis and appveyor.

**Comment by flip1995** [comment]
After adding the `GITHUB_TOKEN` to travis and appveyor, atleast travis will compile and work with the `rustc` master. appveyor will compile with `rustc` master, but the dogfood test fails because of a weird linking error (I couldn't find the culprit for this...).

If no `GITHUB_TOKEN` is set CI will try to use the nightly toolchain.

**Comment by mati865** [comment]
@matthiaskrgr [has found out the reason](https://github.com/rust-lang-nursery/rust-clippy/pull/3072#issuecomment-415200039) for this weird issue:
```
/home/travis/.cargo/bin/clippy-driver: error while loading shared libraries: librustc_driver-b59ab39da03fc19f.so: cannot open shared object file: No such file or directory
```

In other words:
Removing `rust-toolchain` file in CI script could make `export LD_LIBRARY_PATH=$(rustc --print sysroot)/lib` redundant and maybe even fix AppVeyor.

**Comment by flip1995** [comment]
@mati865 we're already doing this in [Travis](https://github.com/rust-lang-nursery/rust-clippy/pull/2941/files#diff-354f30a63fb0907d4ad57269548329e3R55) and in [appveyor](https://github.com/rust-lang-nursery/rust-clippy/pull/2941/files#diff-180360612c6b8c4ed830919bbb4dd459R15).

The appveyor error is somehow connected to the master toolchain. Even when you change nothing in the appveyor script, except the toolchain (from nightly to master) , this error appears. Also adding `$(rustc --print sysroot)/lib` to the path in appveyor didn't fix the problem. I'm all out of ideas on this one...

**Comment by mati865** [comment]
Oops my bad, guess it's the time to get some sleep.

Looking at AppVeyor error again:
```
stderr: error: unable to unlink old fallback exe
info: caused by: Access is denied. (os error 5)
```
It just means the handle to the file is still open. On Windows files cannot be renamed/moved until very last handle is closed.
No idea why this happens with master toolchain.

**Comment by flip1995** [comment]
Do we want to move forward with this? The next steps would be:
- [ ] Create a `GITHUB_TOKEN` and add it to Appveyor and Travis (@oli-obk can you do this? I don't have the permission for Appveyor) [GitHub Doc](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/)
- [ ] Fix dogfood test under windows

**Comment by phansch** [comment]
Just realized that I can add new environment variables on travis, too. I will give it a try now. I don't have access to AppVeyor, though.

**Comment by Manishearth** [comment]
Yeah, just fix it for travis and land this, let me know what we need to do for appveyor and I'll do it. I don't know how to grant access to appveyor.

**Comment by phansch** [comment]
Did a merge instead of a rebase because the `prs` branch was 50+ commits behind master and I'm not sure if any of these commits are used somewhere. Let's see what Travis is saying now.

For AppVeyor, the docs say that there's an 'Environment Variables' tab in the AppVeyor project settings. So essentially, you create a new Token with `public_repo` access [here](https://github.com/settings/tokens/new) and copy the token over to AppVeyor.

**Comment by Manishearth** [comment]
Feel free to merge when ready.

**Comment by phansch** [comment]
Travis looks fine now, but I think AppVeyor is not detecting the `GITHUB_TOKEN` correctly and doesn't install rust from master:
https://ci.appveyor.com/project/rust-lang-libs/rust-clippy/build/1.0.1536#L29
I guess it should show the output of the master installer?

**Comment by Manishearth** [comment]
Yeah not sure what's going on. Feel free to merge and open a new issue for
appveyor, at least we'll have _some_ reliable CI.

**Comment by flip1995** [comment]
It seems that no env var `GITHUB_TOKEN` is defined in the Appveyor settings. It worked on my fork setup: https://ci.appveyor.com/project/flip1995/rust-clippy/build/1.0.28
image

@Manishearth Is there maybe a typo in the env var name? Lower case, instead of upper case? 🤔

**Comment by Manishearth** [comment]
Oh, I added it to Manishearth/rust.

I don't have access to rust-lang-nursery, but I pinged nrc to get it added.

Relatedly: What should we do for PRs then? They don't have access to these env vars, right? Should we use a different method for getting the master hash?

**Comment by kennytm** [comment]
You could clone the rust-lang/rust repo and find the (full 40 digits) hash. I think `git clone` isn't rate limited by IP (otherwise these CI services simply won't work). Cloning would be slower though.

**Comment by Manishearth** [comment]
Yeah, I _really_ want to avoid that but it might be the only option

**Comment by Manishearth** [comment]
Anyway, merging this since at least master gets good CI

**Comment by Manishearth** [comment]
I'm going to add a CGI script for this on my server so we don't need the token.

**Comment by Manishearth** [comment]
Yeah, shallow cloning rust is the other option but i thought it was terrible amd maybe slow so didn't pick it. But really both suck, i don't know which to pick. I guess cloning is nicer since it's less infra I have to maintain.

**Comment by Manishearth** [comment]
`git ls-remote https://github.com/rust-lang/rust.git master` with some
parsing works.

I'll incorporate this into this PR (we can also make this part of
rustup-toolchain-install-master)

**Pull request merged**